src/neuron/cache/mechanism_range.hpp

```diff
@@ -0,0 +1,256 @@
+#pragma once
+#include "membfunc.h"
+#include "nrn_ansi.h"
+#include "nrnoc_ml.h"
+
+#include <array>
+
+namespace neuron::cache {
+/**
+ * @brief Call the given method with each dparam index that should be cached for a mechanism.
+ *
+ * The callable will be invoked with the largest index first. This is useful if you want to resize a
+ * container and use the indices as offsets into it.
+ */
+template <typename Callable>
+void indices_to_cache(short type, Callable callable) {
+    auto const pdata_size = nrn_prop_dparam_size_[type];
+    auto* const dparam_semantics = memb_func[type].dparam_semantics;
+    for (int field = pdata_size - 1; field >= 0; --field) {
+        // Check if the field-th dparam of this mechanism type is an ion variable. See
+        // hoc_register_dparam_semantics.
+        auto const sem = dparam_semantics[field];
+        // TODO is area constant? could we cache the value instead of a pointer to it?
+        // TODO ion type can be handled differently
+        if ((sem > 0 && sem < 1000) || sem == -1 /* area */) {
+            std::invoke(callable, field);
+        }
+    }
+}
+
+/**
+ * @brief Version of Memb_list for use in performance-critical code.
+ *
+ * Unlike Memb_list, this requires that the number of fields is known at compile time.
+ * This is typically only true in translated MOD file code. The idea is that an instance of this
+ * class will be created outside a loop over the data vectors and then used inside the loop.
+ *
+ * @warning It is the responsibility of the caller to ensure that the model remains sorted beyond
```

Comment:
I think not, it should only be relevant for developers who are touching the code generation.
I did try to avoid the limitation/danger at some point, but the cost/overhead didn't seem justified.